Add checks for gpu-kernel calling conv

Flakebi opened the pull request:
The `gpu-kernel` calling convention has several restrictions that were not enforced by the compiler until now.
Add the following restrictions:

1. Cannot be async
2. Cannot be called
3. Cannot return values, return type must be `()` or `!`
4. Arguments should be simple, i.e. passed by value. More complicated types can work when you know what you are doing, but it is rather unintuitive, one needs to know ABI/compiler internals.
5. Export name should be unmangled, either through `no_mangle` or `export_name`. Kernels are searched by name on the CPU side, having a mangled name makes it hard to find and probably almost always unintentional.

Tracking issue: rust-lang/rust#135467
amdgpu target tracking issue: rust-lang/rust#135024

@workingjubilee, these should be all the restrictions we talked about a year ago.

cc @RDambrosio016 @kjetilkjeka for nvptx
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

rust-log-analyzer commented:
The job **`x86_64-gnu-gcc`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/20209916394/job/58014214939) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/58014214939) [(plain)](https://github.com/rust-lang/rust/commit/1e9b1dceea0796d183e0421f5f4bda3245693158/checks/58014214939/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [ui] tests/ui/abi/cannot-return.rs#amdgpu stdout ----

error in revision `amdgpu`: auxiliary build of /checkout/tests/auxiliary/minicore.rs failed to compile: 
status: exit status: 101
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/auxiliary/minicore.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "-Zcodegen-backend=gcc" "--cfg" "amdgpu" "--check-cfg" "cfg(test,FALSE,amdgpu,nvptx)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/abi/cannot-return.amdgpu/libminicore.rlib" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Cpanic=abort" "-Cforce-unwind-tables=yes" "--edition=2024" "--target" "amdgcn-amd-amdhsa" "-Ctarget-cpu=gfx900" "--crate-type=rlib" "--crate-type" "rlib" "-Cpanic=abort"
stdout: none
--- stderr -------------------------------

thread 'rustc' (86344) panicked at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/gccjit-3.1.1/src/lib.rs:119:17:
libgccjit needs to be loaded by calling load() before attempting to do any operation
---
note: please make sure that you have updated to the latest nightly

note: please attach the file at `/checkout/obj/rustc-ice-2025-12-14T16_03_13-86336.txt` to your bug report

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -Z ignore-directory-in-diagnostics-source-blocks=/checkout/vendor -Z codegen-backend=gcc -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0 -C panic=abort -C force-unwind-tables=yes -C target-cpu=gfx900 --crate-type rlib --crate-type rlib -C panic=abort

query stack during panic:
end of query stack
------------------------------------------

---- [ui] tests/ui/abi/cannot-return.rs#amdgpu stdout end ----
---- [ui] tests/ui/abi/cannot-return.rs#nvptx stdout ----

error in revision `nvptx`: auxiliary build of /checkout/tests/auxiliary/minicore.rs failed to compile: 
status: exit status: 101
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/auxiliary/minicore.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "-Zcodegen-backend=gcc" "--cfg" "nvptx" "--check-cfg" "cfg(test,FALSE,amdgpu,nvptx)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/abi/cannot-return.nvptx/libminicore.rlib" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Cpanic=abort" "-Cforce-unwind-tables=yes" "--edition=2024" "--target" "nvptx64-nvidia-cuda" "--crate-type=rlib" "--crate-type" "rlib" "-Cpanic=abort"
stdout: none
--- stderr -------------------------------

thread 'rustc' (86360) panicked at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/gccjit-3.1.1/src/lib.rs:119:17:
libgccjit needs to be loaded by calling load() before attempting to do any operation
---
note: please make sure that you have updated to the latest nightly

note: please attach the file at `/checkout/obj/rustc-ice-2025-12-14T16_03_13-86352.txt` to your bug report

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -Z ignore-directory-in-diagnostics-source-blocks=/checkout/vendor -Z codegen-backend=gcc -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0 -C panic=abort -C force-unwind-tables=yes --crate-type rlib --crate-type rlib -C panic=abort

query stack during panic:
end of query stack
------------------------------------------

---- [ui] tests/ui/abi/cannot-return.rs#nvptx stdout end ----
---- [ui] tests/ui/lint/lint-gpu-kernel.rs#amdgpu stdout ----

error in revision `amdgpu`: auxiliary build of /checkout/tests/auxiliary/minicore.rs failed to compile: 
status: exit status: 101
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/auxiliary/minicore.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "-Zcodegen-backend=gcc" "--cfg" "amdgpu" "--check-cfg" "cfg(test,FALSE,amdgpu,nvptx)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/lint-gpu-kernel.amdgpu/libminicore.rlib" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Cpanic=abort" "-Cforce-unwind-tables=yes" "--edition=2024" "--target" "amdgcn-amd-amdhsa" "-Ctarget-cpu=gfx900" "--crate-type" "rlib" "-Cpanic=abort"
stdout: none
--- stderr -------------------------------

thread 'rustc' (176438) panicked at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/gccjit-3.1.1/src/lib.rs:119:17:
libgccjit needs to be loaded by calling load() before attempting to do any operation
---
note: please make sure that you have updated to the latest nightly

note: please attach the file at `/checkout/obj/rustc-ice-2025-12-14T16_08_49-176429.txt` to your bug report

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -Z ignore-directory-in-diagnostics-source-blocks=/checkout/vendor -Z codegen-backend=gcc -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0 -C panic=abort -C force-unwind-tables=yes -C target-cpu=gfx900 --crate-type rlib -C panic=abort

query stack during panic:
end of query stack
------------------------------------------

---- [ui] tests/ui/lint/lint-gpu-kernel.rs#amdgpu stdout end ----
---- [ui] tests/ui/lint/lint-gpu-kernel.rs#nvptx stdout ----

error in revision `nvptx`: auxiliary build of /checkout/tests/auxiliary/minicore.rs failed to compile: 
status: exit status: 101
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/auxiliary/minicore.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "-Zcodegen-backend=gcc" "--cfg" "nvptx" "--check-cfg" "cfg(test,FALSE,amdgpu,nvptx)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/lint/lint-gpu-kernel.nvptx/libminicore.rlib" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-Cpanic=abort" "-Cforce-unwind-tables=yes" "--edition=2024" "--target" "nvptx64-nvidia-cuda" "--crate-type" "rlib" "-Cpanic=abort"
stdout: none
--- stderr -------------------------------

thread 'rustc' (176447) panicked at /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/gccjit-3.1.1/src/lib.rs:119:17:
libgccjit needs to be loaded by calling load() before attempting to do any operation
---
note: please make sure that you have updated to the latest nightly

note: please attach the file at `/checkout/obj/rustc-ice-2025-12-14T16_08_49-176444.txt` to your bug report

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -Z ignore-directory-in-diagnostics-source-blocks=/checkout/vendor -Z codegen-backend=gcc -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0 -C panic=abort -C force-unwind-tables=yes --crate-type rlib -C panic=abort

query stack during panic:
end of query stack
------------------------------------------

---- [ui] tests/ui/lint/lint-gpu-kernel.rs#nvptx stdout end ----

failures:
    [ui] tests/ui/abi/cannot-return.rs#amdgpu
    [ui] tests/ui/abi/cannot-return.rs#nvptx
    [ui] tests/ui/lint/lint-gpu-kernel.rs#amdgpu
```

</details>

For more information how to resolve CI failures of this job, visit this [link](https://rustc-dev-guide.rust-lang.org/tests/codegen-backend-tests/cg_gcc.html).

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/20213855554/job/58023744405) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/58023744405) [(plain)](https://github.com/rust-lang/rust/commit/10b32d6b5709228906def417ddbdbe15620c7c9b/checks/58023744405/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [ui] tests/ui/zero-sized/zero-sized-btreemap-insert.rs ... ok

failures:

---- [ui] tests/ui/abi/cannot-return.rs#amdgpu stdout ----
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/abi/cannot-return.amdgpu/cannot-return.amdgpu.stderr`
diff of stderr:

1 error: invalid signature for `extern "gpu-kernel"` function
-   --> $DIR/cannot-return.rs:16:37
+   --> $DIR/cannot-return.rs:17:37
3    |
4 LL | extern "gpu-kernel" fn ret_i32() -> i32 { 0 }
5    |                                     ^^^

6    |
7    = note: functions with the "gpu-kernel" ABI cannot have a return type
8 help: remove the return type
-   --> $DIR/cannot-return.rs:16:37
+   --> $DIR/cannot-return.rs:17:37
10    |
11 LL | extern "gpu-kernel" fn ret_i32() -> i32 { 0 }
12    |                                     ^^^

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/abi/cannot-return.rs:17:37
-   --> /checkout/tests/ui/abi/cannot-return.rs:17:37
---
The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args abi/cannot-return.rs`

error in revision `amdgpu`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/abi/cannot-return.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--cfg" "amdgpu" "--check-cfg" "cfg(test,FALSE,amdgpu,nvptx)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/abi/cannot-return.amdgpu" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-Cpanic=abort" "-Cforce-unwind-tables=yes" "--edition=2024" "--target" "amdgcn-amd-amdhsa" "-Ctarget-cpu=gfx900" "--crate-type=rlib" "--extern" "minicore=/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/abi/cannot-return.amdgpu/libminicore.rlib"
stdout: none
--- stderr -------------------------------
error: invalid signature for `extern "gpu-kernel"` function
##[error]  --> /checkout/tests/ui/abi/cannot-return.rs:17:37
   |
LL | extern "gpu-kernel" fn ret_i32() -> i32 { 0 }
   |                                     ^^^
   |
   = note: functions with the "gpu-kernel" ABI cannot have a return type
help: remove the return type
  --> /checkout/tests/ui/abi/cannot-return.rs:17:37
   |
LL | extern "gpu-kernel" fn ret_i32() -> i32 { 0 }
   |                                     ^^^

error: aborting due to 1 previous error
------------------------------------------

---- [ui] tests/ui/abi/cannot-return.rs#amdgpu stdout end ----
---- [ui] tests/ui/abi/cannot-return.rs#nvptx stdout ----
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/abi/cannot-return.nvptx/cannot-return.nvptx.stderr`
diff of stderr:

1 error: invalid signature for `extern "gpu-kernel"` function
-   --> $DIR/cannot-return.rs:16:37
+   --> $DIR/cannot-return.rs:17:37
3    |
4 LL | extern "gpu-kernel" fn ret_i32() -> i32 { 0 }
5    |                                     ^^^

6    |
7    = note: functions with the "gpu-kernel" ABI cannot have a return type
8 help: remove the return type
-   --> $DIR/cannot-return.rs:16:37
+   --> $DIR/cannot-return.rs:17:37
10    |
11 LL | extern "gpu-kernel" fn ret_i32() -> i32 { 0 }
12    |                                     ^^^

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/abi/cannot-return.rs:17:37
-   --> /checkout/tests/ui/abi/cannot-return.rs:17:37
---
The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args abi/cannot-return.rs`

error in revision `nvptx`: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/abi/cannot-return.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--cfg" "nvptx" "--check-cfg" "cfg(test,FALSE,amdgpu,nvptx)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/abi/cannot-return.nvptx" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-Cpanic=abort" "-Cforce-unwind-tables=yes" "--edition=2024" "--target" "nvptx64-nvidia-cuda" "--crate-type=rlib" "--extern" "minicore=/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/abi/cannot-return.nvptx/libminicore.rlib"
stdout: none
--- stderr -------------------------------
error: invalid signature for `extern "gpu-kernel"` function
##[error]  --> /checkout/tests/ui/abi/cannot-return.rs:17:37
   |
LL | extern "gpu-kernel" fn ret_i32() -> i32 { 0 }
   |                                     ^^^
   |
   = note: functions with the "gpu-kernel" ABI cannot have a return type
help: remove the return type
  --> /checkout/tests/ui/abi/cannot-return.rs:17:37
   |
LL | extern "gpu-kernel" fn ret_i32() -> i32 { 0 }
   |                                     ^^^

error: aborting due to 1 previous error
------------------------------------------

---- [ui] tests/ui/abi/cannot-return.rs#nvptx stdout end ----
---- [ui] tests/ui/lint/lint-gpu-kernel.rs#amdgpu stdout ----
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lint/lint-gpu-kernel.amdgpu/lint-gpu-kernel.amdgpu.stderr`
diff of stderr:

1 warning: `extern` fn uses type `()`, which is not FFI-safe
-   --> $DIR/lint-gpu-kernel.rs:35:35
+   --> $DIR/lint-gpu-kernel.rs:36:35
3    |
4 LL | extern "gpu-kernel" fn arg_zst(_: ()) { }
5    |                                   ^^ not FFI-safe

9    = note: `#[warn(improper_ctypes_definitions)]` on by default
10 
11 warning: passing type `()` to a function with "gpu-kernel" ABI may have unexpected behavior
-   --> $DIR/lint-gpu-kernel.rs:35:35
+   --> $DIR/lint-gpu-kernel.rs:36:35
13    |
14 LL | extern "gpu-kernel" fn arg_zst(_: ()) { }
15    |                                   ^^

18    = note: `#[warn(improper_gpu_kernel_arg)]` on by default
19 
20 warning: passing type `&i32` to a function with "gpu-kernel" ABI may have unexpected behavior
-   --> $DIR/lint-gpu-kernel.rs:40:35
+   --> $DIR/lint-gpu-kernel.rs:41:35
22    |
23 LL | extern "gpu-kernel" fn arg_ref(_: &i32) { }
24    |                                   ^^^^

26    = help: use primitive types and raw pointers to get reliable behavior
27 
28 warning: passing type `&mut i32` to a function with "gpu-kernel" ABI may have unexpected behavior
-   --> $DIR/lint-gpu-kernel.rs:43:39
+   --> $DIR/lint-gpu-kernel.rs:44:39
30    |
31 LL | extern "gpu-kernel" fn arg_ref_mut(_: &mut i32) { }
32    |                                       ^^^^^^^^

34    = help: use primitive types and raw pointers to get reliable behavior
35 
36 warning: `extern` fn uses type `S`, which is not FFI-safe
-   --> $DIR/lint-gpu-kernel.rs:48:38
+   --> $DIR/lint-gpu-kernel.rs:49:38
38    |
39 LL | extern "gpu-kernel" fn arg_struct(_: S) { }
40    |                                      ^ not FFI-safe

42    = help: consider adding a `#[repr(C)]` or `#[repr(transparent)]` attribute to this struct
43    = note: this struct has unspecified layout
44 note: the type is defined here
-   --> $DIR/lint-gpu-kernel.rs:46:1
+   --> $DIR/lint-gpu-kernel.rs:47:1
46    |
47 LL | struct S { a: i32, b: i32 }
48    | ^^^^^^^^

49 
50 warning: passing type `S` to a function with "gpu-kernel" ABI may have unexpected behavior
-   --> $DIR/lint-gpu-kernel.rs:48:38
+   --> $DIR/lint-gpu-kernel.rs:49:38
52    |
53 LL | extern "gpu-kernel" fn arg_struct(_: S) { }
54    |                                      ^

56    = help: use primitive types and raw pointers to get reliable behavior
57 
58 warning: `extern` fn uses type `(i32, i32)`, which is not FFI-safe
-   --> $DIR/lint-gpu-kernel.rs:53:35
+   --> $DIR/lint-gpu-kernel.rs:54:35
60    |
61 LL | extern "gpu-kernel" fn arg_tup(_: (i32, i32)) { }
62    |                                   ^^^^^^^^^^ not FFI-safe

65    = note: tuples have unspecified layout
66 
67 warning: passing type `(i32, i32)` to a function with "gpu-kernel" ABI may have unexpected behavior
-   --> $DIR/lint-gpu-kernel.rs:53:35
+   --> $DIR/lint-gpu-kernel.rs:54:35
69    |
70 LL | extern "gpu-kernel" fn arg_tup(_: (i32, i32)) { }
71    |                                   ^^^^^^^^^^

73    = help: use primitive types and raw pointers to get reliable behavior
74 
75 warning: function with the "gpu-kernel" ABI has a mangled name
-   --> $DIR/lint-gpu-kernel.rs:60:1
+   --> $DIR/lint-gpu-kernel.rs:61:1
77    |
78 LL | pub extern "gpu-kernel" fn mangled_kernel() { }
79    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:36:35
-   --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:36:35
---
The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lint/lint-gpu-kernel.rs`

error in revision `amdgpu`: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/lint-gpu-kernel.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--cfg" "amdgpu" "--check-cfg" "cfg(test,FALSE,amdgpu,nvptx)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lint/lint-gpu-kernel.amdgpu" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-Cpanic=abort" "-Cforce-unwind-tables=yes" "--edition=2024" "--target" "amdgcn-amd-amdhsa" "-Ctarget-cpu=gfx900" "--extern" "minicore=/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lint/lint-gpu-kernel.amdgpu/libminicore.rlib"
stdout: none
--- stderr -------------------------------
warning: `extern` fn uses type `()`, which is not FFI-safe
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:36:35
   |
LL | extern "gpu-kernel" fn arg_zst(_: ()) { }
   |                                   ^^ not FFI-safe
   |
   = help: consider using a struct instead
   = note: tuples have unspecified layout
   = note: `#[warn(improper_ctypes_definitions)]` on by default

warning: passing type `()` to a function with "gpu-kernel" ABI may have unexpected behavior
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:36:35
   |
LL | extern "gpu-kernel" fn arg_zst(_: ()) { }
   |                                   ^^
   |
   = help: use primitive types and raw pointers to get reliable behavior
   = note: `#[warn(improper_gpu_kernel_arg)]` on by default

warning: passing type `&i32` to a function with "gpu-kernel" ABI may have unexpected behavior
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:41:35
   |
LL | extern "gpu-kernel" fn arg_ref(_: &i32) { }
   |                                   ^^^^
   |
   = help: use primitive types and raw pointers to get reliable behavior

warning: passing type `&mut i32` to a function with "gpu-kernel" ABI may have unexpected behavior
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:44:39
   |
LL | extern "gpu-kernel" fn arg_ref_mut(_: &mut i32) { }
   |                                       ^^^^^^^^
   |
   = help: use primitive types and raw pointers to get reliable behavior

warning: `extern` fn uses type `S`, which is not FFI-safe
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:49:38
   |
LL | extern "gpu-kernel" fn arg_struct(_: S) { }
   |                                      ^ not FFI-safe
   |
   = help: consider adding a `#[repr(C)]` or `#[repr(transparent)]` attribute to this struct
   = note: this struct has unspecified layout
note: the type is defined here
  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:47:1
   |
LL | struct S { a: i32, b: i32 }
   | ^^^^^^^^

warning: passing type `S` to a function with "gpu-kernel" ABI may have unexpected behavior
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:49:38
   |
LL | extern "gpu-kernel" fn arg_struct(_: S) { }
   |                                      ^
   |
   = help: use primitive types and raw pointers to get reliable behavior

warning: `extern` fn uses type `(i32, i32)`, which is not FFI-safe
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:54:35
   |
LL | extern "gpu-kernel" fn arg_tup(_: (i32, i32)) { }
   |                                   ^^^^^^^^^^ not FFI-safe
   |
   = help: consider using a struct instead
   = note: tuples have unspecified layout

warning: passing type `(i32, i32)` to a function with "gpu-kernel" ABI may have unexpected behavior
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:54:35
   |
LL | extern "gpu-kernel" fn arg_tup(_: (i32, i32)) { }
   |                                   ^^^^^^^^^^
   |
   = help: use primitive types and raw pointers to get reliable behavior

warning: function with the "gpu-kernel" ABI has a mangled name
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:61:1
   |
LL | pub extern "gpu-kernel" fn mangled_kernel() { }
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: use `unsafe(no_mangle)` or `unsafe(export_name = "<name>")`
   = note: mangled names make it hard to find the kernel, this is usually not intended
   = note: `#[warn(missing_gpu_kernel_export_name)]` on by default

warning: 9 warnings emitted
------------------------------------------

---- [ui] tests/ui/lint/lint-gpu-kernel.rs#amdgpu stdout end ----
---- [ui] tests/ui/lint/lint-gpu-kernel.rs#nvptx stdout ----
Saved the actual stderr to `/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lint/lint-gpu-kernel.nvptx/lint-gpu-kernel.nvptx.stderr`
diff of stderr:

1 warning: `extern` fn uses type `()`, which is not FFI-safe
-   --> $DIR/lint-gpu-kernel.rs:35:35
+   --> $DIR/lint-gpu-kernel.rs:36:35
3    |
4 LL | extern "gpu-kernel" fn arg_zst(_: ()) { }
5    |                                   ^^ not FFI-safe

9    = note: `#[warn(improper_ctypes_definitions)]` on by default
10 
11 warning: passing type `()` to a function with "gpu-kernel" ABI may have unexpected behavior
-   --> $DIR/lint-gpu-kernel.rs:35:35
+   --> $DIR/lint-gpu-kernel.rs:36:35
13    |
14 LL | extern "gpu-kernel" fn arg_zst(_: ()) { }
15    |                                   ^^

18    = note: `#[warn(improper_gpu_kernel_arg)]` on by default
19 
20 warning: passing type `&i32` to a function with "gpu-kernel" ABI may have unexpected behavior
-   --> $DIR/lint-gpu-kernel.rs:40:35
+   --> $DIR/lint-gpu-kernel.rs:41:35
22    |
23 LL | extern "gpu-kernel" fn arg_ref(_: &i32) { }
24    |                                   ^^^^

26    = help: use primitive types and raw pointers to get reliable behavior
27 
28 warning: passing type `&mut i32` to a function with "gpu-kernel" ABI may have unexpected behavior
-   --> $DIR/lint-gpu-kernel.rs:43:39
+   --> $DIR/lint-gpu-kernel.rs:44:39
30    |
31 LL | extern "gpu-kernel" fn arg_ref_mut(_: &mut i32) { }
32    |                                       ^^^^^^^^

34    = help: use primitive types and raw pointers to get reliable behavior
35 
36 warning: `extern` fn uses type `S`, which is not FFI-safe
-   --> $DIR/lint-gpu-kernel.rs:48:38
+   --> $DIR/lint-gpu-kernel.rs:49:38
38    |
39 LL | extern "gpu-kernel" fn arg_struct(_: S) { }
40    |                                      ^ not FFI-safe

42    = help: consider adding a `#[repr(C)]` or `#[repr(transparent)]` attribute to this struct
43    = note: this struct has unspecified layout
44 note: the type is defined here
-   --> $DIR/lint-gpu-kernel.rs:46:1
+   --> $DIR/lint-gpu-kernel.rs:47:1
46    |
47 LL | struct S { a: i32, b: i32 }
48    | ^^^^^^^^

49 
50 warning: passing type `S` to a function with "gpu-kernel" ABI may have unexpected behavior
-   --> $DIR/lint-gpu-kernel.rs:48:38
+   --> $DIR/lint-gpu-kernel.rs:49:38
52    |
53 LL | extern "gpu-kernel" fn arg_struct(_: S) { }
54    |                                      ^

56    = help: use primitive types and raw pointers to get reliable behavior
57 
58 warning: `extern` fn uses type `(i32, i32)`, which is not FFI-safe
-   --> $DIR/lint-gpu-kernel.rs:53:35
+   --> $DIR/lint-gpu-kernel.rs:54:35
60    |
61 LL | extern "gpu-kernel" fn arg_tup(_: (i32, i32)) { }
62    |                                   ^^^^^^^^^^ not FFI-safe

65    = note: tuples have unspecified layout
66 
67 warning: passing type `(i32, i32)` to a function with "gpu-kernel" ABI may have unexpected behavior
-   --> $DIR/lint-gpu-kernel.rs:53:35
+   --> $DIR/lint-gpu-kernel.rs:54:35
69    |
70 LL | extern "gpu-kernel" fn arg_tup(_: (i32, i32)) { }
71    |                                   ^^^^^^^^^^

73    = help: use primitive types and raw pointers to get reliable behavior
74 
75 warning: function with the "gpu-kernel" ABI has a mangled name
-   --> $DIR/lint-gpu-kernel.rs:60:1
+   --> $DIR/lint-gpu-kernel.rs:61:1
77    |
78 LL | pub extern "gpu-kernel" fn mangled_kernel() { }
79    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Note: some mismatched output was normalized before being compared
-   --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:36:35
-   --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:36:35
---
The actual stderr differed from the expected stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args lint/lint-gpu-kernel.rs`

error in revision `nvptx`: 1 errors occurred comparing output.
status: exit status: 0
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/lint/lint-gpu-kernel.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2" "--cfg" "nvptx" "--check-cfg" "cfg(test,FALSE,amdgpu,nvptx)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lint/lint-gpu-kernel.nvptx" "-A" "unused" "-W" "unused_attributes" "-A" "internal_features" "-A" "unused_parens" "-A" "unused_braces" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/aarch64-unknown-linux-gnu/native/rust-test-helpers" "-Cpanic=abort" "-Cforce-unwind-tables=yes" "--edition=2024" "--target" "nvptx64-nvidia-cuda" "--extern" "minicore=/checkout/obj/build/aarch64-unknown-linux-gnu/test/ui/lint/lint-gpu-kernel.nvptx/libminicore.rlib"
stdout: none
--- stderr -------------------------------
warning: `extern` fn uses type `()`, which is not FFI-safe
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:36:35
   |
LL | extern "gpu-kernel" fn arg_zst(_: ()) { }
   |                                   ^^ not FFI-safe
   |
   = help: consider using a struct instead
   = note: tuples have unspecified layout
   = note: `#[warn(improper_ctypes_definitions)]` on by default

warning: passing type `()` to a function with "gpu-kernel" ABI may have unexpected behavior
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:36:35
   |
LL | extern "gpu-kernel" fn arg_zst(_: ()) { }
   |                                   ^^
   |
   = help: use primitive types and raw pointers to get reliable behavior
   = note: `#[warn(improper_gpu_kernel_arg)]` on by default

warning: passing type `&i32` to a function with "gpu-kernel" ABI may have unexpected behavior
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:41:35
   |
LL | extern "gpu-kernel" fn arg_ref(_: &i32) { }
   |                                   ^^^^
   |
   = help: use primitive types and raw pointers to get reliable behavior

warning: passing type `&mut i32` to a function with "gpu-kernel" ABI may have unexpected behavior
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:44:39
   |
LL | extern "gpu-kernel" fn arg_ref_mut(_: &mut i32) { }
   |                                       ^^^^^^^^
   |
   = help: use primitive types and raw pointers to get reliable behavior

warning: `extern` fn uses type `S`, which is not FFI-safe
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:49:38
   |
LL | extern "gpu-kernel" fn arg_struct(_: S) { }
   |                                      ^ not FFI-safe
   |
   = help: consider adding a `#[repr(C)]` or `#[repr(transparent)]` attribute to this struct
   = note: this struct has unspecified layout
note: the type is defined here
  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:47:1
   |
LL | struct S { a: i32, b: i32 }
   | ^^^^^^^^

warning: passing type `S` to a function with "gpu-kernel" ABI may have unexpected behavior
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:49:38
   |
LL | extern "gpu-kernel" fn arg_struct(_: S) { }
   |                                      ^
   |
   = help: use primitive types and raw pointers to get reliable behavior

warning: `extern` fn uses type `(i32, i32)`, which is not FFI-safe
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:54:35
   |
LL | extern "gpu-kernel" fn arg_tup(_: (i32, i32)) { }
   |                                   ^^^^^^^^^^ not FFI-safe
   |
   = help: consider using a struct instead
   = note: tuples have unspecified layout

warning: passing type `(i32, i32)` to a function with "gpu-kernel" ABI may have unexpected behavior
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:54:35
   |
LL | extern "gpu-kernel" fn arg_tup(_: (i32, i32)) { }
   |                                   ^^^^^^^^^^
   |
   = help: use primitive types and raw pointers to get reliable behavior

warning: function with the "gpu-kernel" ABI has a mangled name
##[warning]  --> /checkout/tests/ui/lint/lint-gpu-kernel.rs:61:1
   |
LL | pub extern "gpu-kernel" fn mangled_kernel() { }
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: use `unsafe(no_mangle)` or `unsafe(export_name = "<name>")`
   = note: mangled names make it hard to find the kernel, this is usually not intended
   = note: `#[warn(missing_gpu_kernel_export_name)]` on by default

warning: 9 warnings emitted
------------------------------------------

---- [ui] tests/ui/lint/lint-gpu-kernel.rs#nvptx stdout end ----

failures:
    [ui] tests/ui/abi/cannot-return.rs#amdgpu
    [ui] tests/ui/abi/cannot-return.rs#nvptx
    [ui] tests/ui/lint/lint-gpu-kernel.rs#amdgpu
```

</details>

workingjubilee commented on compiler/rustc_lint/src/gpukernel_abi.rs:
```diff
@@ -0,0 +1,177 @@
+    /// This lint is issued when it detects a probable mistake in a signature.
+    IMPROPER_GPU_KERNEL_ARG,
+    Warn,
+    "simple arguments of gpu-kernel functions"
```
This line should capture the _reason_ for the lint, not what it is checking for, so something like
```suggestion
    "GPU kernel entry points have a limited calling convention"
```

workingjubilee reviewed (changes requested):
The AST-level code looks good.

Some details on messaging here. I'm not committed to a precise message on these, which is why it's a bit "multiple choice" here, just wondering if these could be improved. In one or two cases it is a must-change.

Should we be enforcing a maximum number of arguments, also? Probably not if there's no cross-driver consensus on that, but maybe?

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149991/693f365667a97b24cb40173bc2801eb66ea53020..88ad16cb286ed4ef33901e2b8c56c082f16349a5)*

Flakebi commented on compiler/rustc_hir_typeck/messages.ftl:
```diff
@@ -133,6 +133,10 @@ hir_typeck_fru_suggestion =
 
+hir_typeck_gpu_kernel_abi_cannot_be_called =
+    functions with the "gpu-kernel" ABI cannot be called
+    .note = an `extern "gpu-kernel"` function can only be launched on the GPU through an API
```
It’s vague because there’s so many APIs :smile:
A gpu-kernel must be loaded and launched through cuda/hip/sycl/hsa or any wrapper around these APIs.
Driver is often used in the graphics space, but I don’t read it often when it comes to GPGPU.
The kernel loader is also part of cuda/hip/…, but not exactly the piece that launches (as in, for launching a process on the CPU, one doesn’t call ld, but the kernel (even though ld is involved)).

The `must` is definitely better, I’ll also remove the `only` (so, `a gpu-kernel function must be launched on the GPU…`).

I think the distinctive part is, only the (GPU) hardware is allowed to start gpu-kernels, they must not be called.
Note: Launching a kernel is not really a call, it’s 1-to-n – launching n instances of the kernel – and it’s async/doesn’t return.

kjetilkjeka replied:
I would suggest `must be launched on the GPU by the runtime`. 

I think that's somewhat standard even though it's both quite a big term, and there's a slight ambiguity related to CUDA calling something the "Driver API" as opposed to the "Runtime API".

workingjubilee replied:
Yes, that description feels like the essence of what I was grasping towards, thank you for muddling through it with me. The big idea here being to clearly say "by someone else that is both *not you* and not really beholden to you, either".

Flakebi commented on compiler/rustc_lint/src/gpukernel_abi.rs:
```diff
@@ -0,0 +1,177 @@
+}
+
+declare_lint! {
+    /// The `missing_gpu_kernel_export_name` lint detects `gpu-kernel` functions that have a mangled name.
```
Yes, if someone calls their kernel `malloc` and we unmangle it, funny things will happen.

Flakebi reviewed (commented):
Thanks for the review!

Some context on the internal workings:
1. On the CPU side, a program passes arguments to a kernel
2. The “API” takes these arguments and writes them into GPU memory
3. The kernel on the GPU gets a pointer to this memory
4. When the kernel accesses arguments, it reads from this memory

(I think nvidia and amd work the same here. I’m not too familiar with nvidia, but this seems to suggest so: https://github.com/Rust-GPU/rust-cuda/blob/44c44baf6fb738d5ffec25aac5db8af02514e890/crates/rustc_codegen_nvvm/src/abi.rs#L60)

So, number of arguments or size of arguments doesn’t really matter, it’s all memory anyways.
And, we could make struct arguments work (maybe, I didn’t look into the details), but Rust would need to take them by value, currently it changes them to pass by pointer.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149991/68f11a11b666b36137374b6e54fc0d013a3b26cf..01f1e1b7c8b124da8f9ad82224937ba7cf93d516)*

workingjubilee commented on compiler/rustc_lint/src/gpukernel_abi.rs:
```diff
@@ -0,0 +1,177 @@
+                | ty::Slice(_)
+                | ty::Str
+                | ty::Tuple(_)
+                | ty::UnsafeBinder(_) => false,
```
The answer for `ty::UnsafeBinder` should be the recursive answer to this question by default (I think the implied result is quite useless given the answer to other things, but that's fine).

workingjubilee commented on compiler/rustc_lint/src/gpukernel_abi.rs:
```diff
@@ -0,0 +1,177 @@
+                | ty::CoroutineClosure(_, _)
+                | ty::CoroutineWitness(..)
+                | ty::Dynamic(_, _)
+                | ty::Error(_)
```
We should cease inspecting the type or just `continue` if the type is erroneous.

workingjubilee reviewed (commented):
> Some context on the internal workings:
>    1. On the CPU side, a program passes arguments to a kernel
>    1. The “API” takes these arguments and writes them into GPU memory
>    1. The kernel on the GPU gets a pointer to this memory
>    1. When the kernel accesses arguments, it reads from this memory

Ah, yeah. I know of the general idea here, though I am foggy on specifics in many cases.

And even in some hypothetical where the driver isn't writing them into the GPU memory, it would have to invoke some dark magic on the GPU to put it directly into registers anyways, which is just saying "did you know: some computer hardware has memory-mapped registers?"

( I only am making that comment because I vaguely remember one GPU driver involving something similar. )

At some point we are left with doing codegen to accept arguments and we can expect that to have some target-specific nuances, even if it's just stack alignment.

For structs, I'd rather we avoid thinking too hard about the struct question for cases where they aren't just `repr(transparent)` primitives until we have tinkered with things a bit more. If `repr(C)` structs are widely supported and fully intentionally that could be worth tackling.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149991/d0835adc4e114eac911150b3692b830b5583223a..6868f66e988848a7f7eeae345ce845f4a2b227b6)*

Flakebi committed:
Add checks for gpu-kernel calling conv

The `gpu-kernel` calling convention has several restrictions that were
not enforced by the compiler until now.
Add the following restrictions:

1. Cannot be async
2. Cannot be called
3. Cannot return values, return type must be `()` or `!`
4. Arguments should be primitives, i.e. passed by value. More complicated
   types can work when you know what you are doing, but it is rather
   unintuitive, one needs to know ABI/compiler internals.
5. Export name should be unmangled, either through `no_mangle` or
   `export_name`. Kernels are searched by name on the CPU side, having
   a mangled name makes it hard to find and probably almost always
   unintentional.

Flakebi commented:
I tried rewriting the lint pass using `TypeFolder` (plus fixing the other comments).

workingjubilee commented:
Cool!

Alright. This looks correct to me. I feel slightly iffy about the linting in general... I feel like we're gonna have some edge-case bite our nose, here... but it's just a lint, so it's okay if it's wrong. The ABI isn't stable, and the things we affirmatively reject are a conservative set of things we know are definitely wrong.

@bors r+
